Added operational model for pthread functions

lucasccordeiro opened the pull request:
Added operational models for these pthread functions: pthread_key_create, pthread_getspecific, and pthread_setspecific.

lucasccordeiro committed:
Added operational model for pthread functions

Added operational models for these pthread functions:
pthread_key_create, pthread_key_delete,
pthread_getspecific, and pthread_setspecific.

feliperodri commented:
This PR solves https://github.com/esbmc/esbmc/issues/35

feliperodri commented on regression/esbmc/01_pthread48/main.c:
```diff
@@ -0,0 +1,46 @@
+/* Contributed by: Vladimír Štill, https://divine.fi.muni.cz
+   Description: A test case for pthread TLS.
+*/
```
Do we have a proper header format? If we do, we should use it; otherwise:
```c
/*
 * Contributed by Vladimír Štill (https://divine.fi.muni.cz).
 * Description: A test case for pthread TLS.
 */
```

lucasccordeiro replied:
Fixed

feliperodri commented on regression/esbmc/01_pthread48/main.c:
```diff
@@ -0,0 +1,46 @@
+
+int main() {
+    pthread_key_t key;
+    int r = pthread_key_create( &key, NULL );
```
You could nondeterministically assign `&key` or `NULL`, so we can explore more paths of the model (check resiliency).
```c
int r = pthread_key_create( nondet_bool() ? &key : NULL, NULL );
```
This could be apply for all the remaining `pthread_*` functions invocations with `&key` as parameter (since test them is the goal of this test).

lucasccordeiro replied:
We expect the first parameter to be a `pthread_key_t`. If we assign `NULL`, then we get `dereference failure: NULL pointer`. I think this suggestion should lead to another test case. 

@mikhailramalho and @feliperodri: What do you think?

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread48/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+
+^VERIFICATION SUCCESSFUL$
```
We should also add a test case that fails, so we can also catch spurious correct results.

lucasccordeiro replied:
Fixed.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +483,39 @@ __ESBMC_HIDE:;
+__attribute__((annotate("__ESBMC_inf_size"))) 
+void (*__ESBMC_thread_key_dtors[])(void *);
+
+unsigned long __ESBMC_next_thread_key;
```
I'd add all the global variables at the top of the file. Also,
`s/__ESBMC_thread_key_dtors/__ESBMC_thread_key_destructors/g`.

lucasccordeiro replied:
Fixed.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +483,39 @@ __ESBMC_HIDE:;
 {
 __ESBMC_HIDE:;
-  // TODO
+  __ESBMC_thread_keys[__ESBMC_next_thread_key] = 0;
```
We should check for nullness here.
```c
__ESBMC_assert(key != NULL, "In pthread_key_create, key parameter should be different than null.");
/* optional destructor function; NULL if not needed */
```

Also, I'd recommend the following for clarity
```c
 __ESBMC_thread_keys[__ESBMC_next_thread_key] = NULL;
```

lucasccordeiro replied:
Fixed.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +483,39 @@ __ESBMC_HIDE:;
+  __ESBMC_thread_keys[__ESBMC_next_thread_key] = 0;
+  __ESBMC_thread_key_dtors[__ESBMC_next_thread_key] = destructor;
+  *key = __ESBMC_next_thread_key++;
+  return 0;
```
We should define the following macros (see [documentation](https://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html))
```c
#define EAGAIN          11      /* Try again */
#define ENOMEM          12      /* Out of memory */
```
and then return them, nondeterministically (see [pthread_key_create documentation](https://linux.die.net/man/3/pthread_key_create)); unless we explicitly assume otherwise (then we must explain our assumptions in the function header). 
```c
return nondet_bool() ? (nondet_bool()? EAGAIN : ENOMEM) : 0;
```

lucasccordeiro replied:
Given that the SV-COMP benchmarks do not consider these return values, I have just added an assumption.

How should we proceed here? SV-COMP makes various assumptions that are not consistent with the real-life scenario (e.g., `malloc` always succeeds).

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +483,39 @@ __ESBMC_HIDE:;
+  return 0;
+}
+
+int pthread_key_delete(pthread_key_t key)
```
_"Any attempt to use key following the call to pthread_key_delete() results in undefined behavior."_ (see [pthread_key_delete documentation](https://linux.die.net/man/3/pthread_key_delete)). We should make explicitly in the header of this function that we do not consider that behavior.

We could consider that all non-initialized positions in `__ESBMC_thread_keys` (during creation) are an specific (-1) object. Something like that in `pthread_key_create`,
```c
size_t nondet_pos;
__ESBMC_assume(nondet_pos != key);
__ESBMC_thread_keys[nondet_pos] = -1;
```

If this works, we could improve this model:
```c
if(__ESBMC_thread_keys[key] == -1) {
    return EINVAL;
}
__ESBMC_thread_keys[key] = 0;
return 0;
```

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +483,39 @@ __ESBMC_HIDE:;
 __ESBMC_HIDE:;
-  // TODO
-  return NULL;
+  return (void *)__ESBMC_thread_keys[key];
```
Similarly to the previous functions, `pthread_getspecific`could return `EINVAL` if the key is invalid or nondeterministically return `ENOMEM` (see [documentation](https://linux.die.net/man/3/pthread_getspecific)).

lucasccordeiro replied:
Fixed.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +483,39 @@ __ESBMC_HIDE:;
 {
 __ESBMC_HIDE:;
-  // TODO
+  __ESBMC_thread_keys[key] = value;
```
Similarly to the previous functions, `pthread_setspecific`could return `EINVAL` if the key is invalid or nondeterministically return `ENOMEM` (see [documentation](https://linux.die.net/man/3/pthread_setspecific)).

lucasccordeiro replied:
Fixed.

lucasccordeiro committed:
fixed header of 01_pthread48

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
moved all global variables to the top of the file

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added precondition in pthread_key_create

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added assumption to pthread_key_create

We assume the pthread_key_create() function will store
the newly created key value at *key and return zero.

lucasccordeiro committed:
added realistic models for pthread

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added test case to check for pthread_key_create

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
fixed indentation

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho commented:
I think also operations should be atomic.

lucasccordeiro committed:
make the pthread methods atomic

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added test case to check pthread_setspecific

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added test case for key parameter

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
fixed code style

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
applied code style

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added test case for pthread destructors

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

feliperodri commented on regression/esbmc/01_pthread48/main.c:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Extracted from https://github.com/sosy-lab/sv-benchmarks).
```
Remove the last ")", it should be:
```c
Extracted from https://github.com/sosy-lab/sv-benchmarks.
```

lucasccordeiro replied:
fixed.

feliperodri commented on regression/esbmc/01_pthread48/main.c:
```diff
@@ -0,0 +1,52 @@
+
+  return 0;
+}
+
```
Remove extra lines at the end of the files.

feliperodri commented on regression/esbmc/01_pthread49/main.c:
```diff
@@ -0,0 +1,10 @@
+int main() {
+  pthread_key_t key;
+  int r = pthread_key_create( &key, NULL );
+  assert( r == 0);
```
fix format
```c
  assert( r == 0 );
```

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread49/main.c:
```diff
@@ -0,0 +1,10 @@
+  assert( r == 0);
+  return 0;
+}
+
```
extra line

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread50/main.c:
```diff
@@ -0,0 +1,22 @@
+  assert( val == 0);
+
+  r = pthread_setspecific( key, nondet_bool() ? (void *)16 : NULL );
+  assert( r == 0 || r == EINVAL || r == ENOMEM);
```
format again, is there a script we can run to automatically apply code style?

feliperodri commented on regression/esbmc/01_pthread50/main.c:
```diff
@@ -0,0 +1,22 @@
+
+  return 0;
+}
+
```
extra line

feliperodri commented on regression/esbmc/01_pthread50/main.c:
```diff
@@ -0,0 +1,22 @@
+int main() {
+  pthread_key_t key;
+  int r = pthread_key_create( &key, NULL );
+  assert( r == 0 || r == EAGAIN || r == ENOMEM);
```
code style

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread51/main.c:
```diff
@@ -0,0 +1,11 @@
+int main() {
+  pthread_key_t key;
+  int r = pthread_key_create( nondet_bool() ? &key : NULL, NULL );
+  assert( r == 0 || r == EAGAIN || r == ENOMEM);
```
code style

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread51/main.c:
```diff
@@ -0,0 +1,11 @@
+  assert( r == 0 || r == EAGAIN || r == ENOMEM);
+  return 0;
+}
+
```
extra line

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread52/main.c:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Extracted from https://github.com/sosy-lab/sv-benchmarks).
+ * Contributed by Vladimír Štill (https://divine.fi.muni.cz).
+ *  Description: A test case for pthread TLS. This test demonstartes that
```
extra space

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread52/main.c:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Extracted from https://github.com/sosy-lab/sv-benchmarks).
```
remove extra ")"

lucasccordeiro replied:
Fixed.

feliperodri reviewed (approved):
Apart from minor code style glitches, LGTM.

lucasccordeiro committed:
applied code style to test cases

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro commented:
@mikhailramalho: May I ask you to review this PR?

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -486,20 +501,37 @@ __ESBMC_HIDE:;
+  __ESBMC_thread_key_destructors[__ESBMC_next_thread_key] = destructor;
+  *key = __ESBMC_next_thread_key++;
+  __ESBMC_atomic_end();
+  return nondet_bool() ? (nondet_bool() ? EAGAIN : ENOMEM) : 0;
```
The model here is wrong: despite succeeding in creating the key, we still might return errors? 

The [documentation](https://pubs.opengroup.org/onlinepubs/007904875/functions/pthread_key_create.html) says:
```
The pthread_key_create() function shall fail if:

[EAGAIN]
The system lacked the necessary resources to create another thread-specific data key, or the system-imposed limit on the total number of keys per process {PTHREAD_KEYS_MAX} has been exceeded.
[ENOMEM]
Insufficient memory exists to create the key.
```

There are public implementations of this function that we can use as base for our model: [here](https://android.googlesource.com/platform/external/pthreads/+/216eb8153b2455d1e15f8fbf84d4413ddcc40adc/pthread_key_create.c)

We should either:
1. Correctly model all the errors that may happen.
2. Assume it always succeeds, as we do for other models in the pthread lib.

In the case of 1, the code should be something like:
```
  atomic_begin();
  int result = 0;
  pthread_key_t newkey = (pthread_key_t) calloc (1, sizeof (*newkey));

  if (newkey == NULL)
    {
      result = ENOMEM;
    }
  else if (/* Check if we can insert the key in the array */)
    {
      result = EAGAIN;

      free (newkey);
      newkey = NULL;
    }
  else if (destructor != NULL)
    {
      newkey->keyLock = 0;
      newkey->destructor = destructor;
    }

  *key = newkey;

  atomic_end();
  return result;
```

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -486,20 +501,37 @@ __ESBMC_HIDE:;
+    return NULL;
+  else
+    return (void *)__ESBMC_thread_keys[key];
+  __ESBMC_atomic_end();
```
We are returning without closing the atomic scope, this will "bleed" into other contexts.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -486,20 +501,37 @@ __ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+  if(value == NULL)
+    return EINVAL;
+
```
returning without closing the atomic scope.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -486,20 +501,37 @@ __ESBMC_HIDE:;
+  __ESBMC_thread_keys[key] = value;
+
+  __ESBMC_atomic_end();
+  return (nondet_bool() ? ENOMEM : 0);
```
We assigned the value to the key and we still might fail?

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -118,6 +124,15 @@ void pthread_exit(void *retval)
+    __ESBMC_thread_keys[i] = 0;
+    if(__ESBMC_thread_key_destructors[i] && key)
+      __ESBMC_thread_key_destructors[i](key);
+  }
```
Why are we cleaning all the keys here?

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -6,6 +6,11 @@ typedef void *(*__ESBMC_thread_start_func_type)(void *);
 
+__attribute__((annotate("__ESBMC_inf_size"))) const void *__ESBMC_thread_keys[];
+__attribute__((annotate("__ESBMC_inf_size"))) void (
+  *__ESBMC_thread_key_destructors[])(void *);
```
Not your fault but: I hate this code formatting.

mikhailramalho commented on regression/esbmc/01_pthread48/main.c:
```diff
@@ -0,0 +1,53 @@
+{
+  pthread_key_t key;
+  int r = pthread_key_create(&key, NULL);
+  assert(r == 0 || r == EAGAIN || r == ENOMEM);
```
This assertion looks dodgy: we either succeeded with 0 and the program should continue, or we failed and all set/getspecific will return wrong values.

This program is testing that it doesn' t matter if we failed or succeeded, the key will always be there.

lucasccordeiro replied:
Fixed.

mikhailramalho reviewed (commented):
tl;dr: the model and test case should be fixed.

lucasccordeiro committed:
fixed atomic block in pthread's OMs

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -486,20 +501,43 @@ __ESBMC_HIDE:;
+  {
+    __ESBMC_atomic_end(); 
+    return (void *)__ESBMC_thread_keys[key];
+  }
```
I think the following is better:
```
__ESBMC_atomic_begin();
void* result = (key > __ESBMC_next_thread_key) ? NULL : (void *)__ESBMC_thread_keys[key];
__ESBMC_atomic_end();
return result;
```

lucasccordeiro committed:
Assume the pthread models always succeed

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added new test case for destructor

lucasccordeiro committed:
check for insufficient memory in pthread models

lucasccordeiro commented:
We could improve our operational models by using dynamic memory allocation for 'pthread_key_create'; however, ESBMC seems to fail in some simple cases for `calloc` and `realloc`:  https://github.com/esbmc/esbmc/issues/151.

lucasccordeiro committed:
call key destructors after a thread returns

lucasccordeiro committed:
temporary implementation

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added linked list to model pthread_key_create

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
renamed insert and search methods for thread-key

lucasccordeiro commented:
This PR is ready to be reviewed.

lucasccordeiro committed:
updated test desc to prune the exploration

lucasccordeiro committed:
deleted key after calling pthread_exec_key_destructors

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

feliperodri commented on regression/esbmc/01_pthread48/main.c:
```diff
@@ -0,0 +1,70 @@
+#include <errno.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdlib.h>
```
nit: Lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread48/main.c:
```diff
@@ -0,0 +1,70 @@
+  return NULL;
+}
+
+pthread_key_t key;
```
why this should be global?

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread48/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--force-malloc-success --incremental-bmc
```
why do we need `--force-malloc-success`? Could we just handle the cases where malloc might fail in the implementation?

lucasccordeiro replied:
We have created test cases to check then the malloc succeeds (an SV-COMP requirement).

We have also other test cases that check when malloc fails.

feliperodri replied:
Great!

feliperodri commented on regression/esbmc/01_pthread49/main.c:
```diff
@@ -0,0 +1,10 @@
+{
+  pthread_key_t key;
+  int r = pthread_key_create(&key, NULL);
+  assert(r == 1); //r must be 0
```
C-style comments? `/* r must be 0. */`

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread50/main.c:
```diff
@@ -0,0 +1,31 @@
+#include <errno.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdlib.h>
```
nit: lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

lucasccordeiro replied:
Fixed.

feliperodri commented on regression/esbmc/01_pthread51/main.c:
```diff
@@ -0,0 +1,16 @@
+#include <errno.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdlib.h>
```
nit: lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

feliperodri commented on regression/esbmc/01_pthread52/main.c:
```diff
@@ -0,0 +1,55 @@
+#include <pthread.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
```
nit: lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

feliperodri commented on regression/esbmc/01_pthread52/main.c:
```diff
@@ -0,0 +1,55 @@
+void dtor(void *v)
+{
+  long val = (long)v;
+  assert(val != 42); //this assertion must fail
```
C-style comments.

feliperodri commented on regression/esbmc/01_pthread53/main.c:
```diff
@@ -0,0 +1,50 @@
+#include <pthread.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
```
nit: lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

feliperodri commented on regression/esbmc/01_pthread53/main.c:
```diff
@@ -0,0 +1,50 @@
+#include <errno.h>
+#include <stdlib.h>
+
+void dtor(void *v)
```
could you change to a more suggestive name?

feliperodri commented on regression/esbmc/01_pthread54/main.c:
```diff
@@ -0,0 +1,67 @@
+#include <errno.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdlib.h>
```
nit: lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

feliperodri commented on regression/esbmc/01_pthread54/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--unwind 3
```
Why do we need this unwind?

feliperodri commented on regression/esbmc/01_pthread55/main.c:
```diff
@@ -0,0 +1,70 @@
+#include <errno.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdlib.h>
```
nit: lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

feliperodri commented on regression/esbmc/01_pthread55/main.c:
```diff
@@ -0,0 +1,70 @@
+  return NULL;
+}
+
+pthread_key_t key;
```
Why this have to be global?

feliperodri commented on regression/esbmc/01_pthread56/main.c:
```diff
@@ -0,0 +1,31 @@
+#include <errno.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdlib.h>
```
nit: lexicographical order.
```c
#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
```

feliperodri commented on regression/esbmc/01_pthread56/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--force-malloc-success --incremental-bmc
```
Same comment about `--force-malloc-success` here.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -45,6 +49,76 @@ pthread_t __ESBMC_get_thread_id(void);
+    l->next = head;
+  }
+  head = l;
+  return 0;
```
Suggestion:
```c
  l = (__ESBMC_thread_key *)malloc(sizeof(__ESBMC_thread_key*));
  if(l == NULL)
  {
    return -1;
  }
  l->thread = __ESBMC_get_thread_id();
  l->key = key;
  l->value = value;
  l->next = (head == NULL) ? NULL : head;
  head = l;
  return 0;
```

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -45,6 +49,76 @@ pthread_t __ESBMC_get_thread_id(void);
+  if(l == NULL)
+    return 0;
+  else
+    return l->value;
```
```c
return ((l == NULL) ? 0 : l->value);
```
or use `{}`, please

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -65,6 +139,29 @@ void pthread_end_main_hook(void)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+  // At thread exit, if a key value has a non-NULL destructor pointer,
+  // and the thread has a non-NULL value associated with that key,
+  // the value of the key is set to NULL, and then the function pointed to
+  // is called with the previously associated value as its sole argument.
+  // The order of destructor calls is unspecified if more than one destructor
+  // exists for a thread when it exits.
+  // source: https://linux.die.net/man/3/pthread_key_create
```
C-style comments.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +582,90 @@ __ESBMC_HIDE:;
   return 0;
 }
 
+// The pthread_key_create() function shall create a thread-specific
+// data key visible to all threads in the process.
+// source: https://linux.die.net/man/3/pthread_key_create
```
C-style comments.

lucasccordeiro replied:
I have followed the style of the comments in ESBMC.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +582,90 @@ __ESBMC_HIDE:;
-  // TODO
-  return 0;
+  __ESBMC_atomic_begin();
+  int result = 0;
```
Put it after the assertion. If the assertion failed, we created a variable on the stack for nothing.

lucasccordeiro replied:
Fixed.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +582,90 @@ __ESBMC_HIDE:;
+    key != NULL,
+    "In pthread_key_create, key parameter must be different than NULL.");
+  // the value NULL shall be associated with the new key in all active threads
+  result = insert_key_value(mylist, __ESBMC_next_thread_key, NULL);
```
`int result = ...`

lucasccordeiro replied:
Fixed.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +582,90 @@ __ESBMC_HIDE:;
+  __ESBMC_thread_key_destructors[__ESBMC_next_thread_key] = destructor;
+  // store the newly created key value at *key
+  *key = __ESBMC_next_thread_key++;
+  if(result < 0)
```
Add a comment here explaining when the result would be < 0 in this case.

lucasccordeiro replied:
Fixed.

feliperodri commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -483,23 +582,90 @@ __ESBMC_HIDE:;
+    // If no thread-specific data value is associated with key,
+    // then the value NULL shall be returned.
+    result = NULL;
+  }
```
Correct me if I'm wrong, but this branch is completely irrelevant here. if `(key <= __ESBMC_next_thread_key)` is not true, key can only be `> __ESBMC_next_thread_key` and result is already null.

lucasccordeiro replied:
Fixed.

feliperodri reviewed (changes requested):
It looks great for me. I only have a couple of suggestions on the linked list implementation and a few questions on the model. I'd also advocate to avoid completely C++-style comments and keep includes in lexicographical order.

lucasccordeiro committed:
added benchmarks from SV-COMP

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added support to reach_error

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
check for null pointers when deleting a node

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
applied code style

lucasccordeiro committed:
moved declaration after the assert-statement

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
check whether we failed to insert the key into our list

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
removed unnecessary branch in pthread_getspecific

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
removed global variable

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
applied code style

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
improved variable name

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
applied code style

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
refactored the insert_key_value implementation

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro commented:
@rafaelsamenezes: could you please run the code of this branch over the SV-COMP `concurrency` benchmarks?

feliperodri committed:
Rename destructor function

feliperodri committed:
Rename destructor function

feliperodri committed:
Update pthread_lib.c

feliperodri reviewed (approved):
LGTM

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -45,6 +49,68 @@ pthread_t __ESBMC_get_thread_id(void);
+{
+  l = (__ESBMC_thread_key *)malloc(sizeof(__ESBMC_thread_key *));
+  if(l == NULL)
+  {
```
No need for braces in one line `if`s.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -45,6 +49,68 @@ pthread_t __ESBMC_get_thread_id(void);
+{
+  l = head;
+  while(l != NULL && l->thread != __ESBMC_get_thread_id())
+  {
```
No need for braces in one line `while`s.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -45,6 +49,68 @@ pthread_t __ESBMC_get_thread_id(void);
+{
+  __ESBMC_thread_key *tmp;
+  if(head == NULL)
+  {
```
No need for braces in one line `if`s.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -45,6 +49,68 @@ pthread_t __ESBMC_get_thread_id(void);
+  if(head != l)
+  {
+    while(tmp->next != NULL && tmp->next != l)
+    {
```
one line `while`s.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/c2goto/library/pthread_lib.c:
```diff
@@ -45,6 +49,68 @@ pthread_t __ESBMC_get_thread_id(void);
+    }
+    tmp->next = l->next;
+  }
+  else if(l->next != NULL)
```
one line `if`s.

lucasccordeiro replied:
Fixed.

mikhailramalho reviewed (changes requested):
I'm confused about the use of both `mylist` and `head`, can't we use only `head`?

lucasccordeiro committed:
removed unnecessary braces

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
fixed typo

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
removed mylist

lucasccordeiro committed:
added missing file

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

rafaelsamenezes commented:
Concurrency Safety results:

- 2 new correct false
- 2 new incorrect false
- Score went from 566 to 536

Incorrect False:
- https://github.com/sosy-lab/sv-benchmarks/blob/master/c/pthread-divine/one_time_barrier_twice_2t.i
- https://github.com/sosy-lab/sv-benchmarks/blob/master/c/pthread-divine/barrier_2t.i

lucasccordeiro commented:
These two incorrect false has no relation to `pthread_key_create, pthread_getspecific, and pthread_setspecific`.

I'll be open another PR to model the functions `pthread_cond_broadcast` and `pthread_mutex_destroy`.

lucasccordeiro commented:
@rafaelsamenezes: Could you please do the following:

1) Download the SV-COMP benchmarks used in the 2020 competition.

2) Run the ESBMC version available in the `master branch` to obtain our **baseline**.

3) Run the ESBMC version available in the `fix-concurrency` branch.

lucasccordeiro commented:
Once this PR is merged, we should test this one https://github.com/esbmc/esbmc/pull/199, which fixes the incorrect false in:

- https://github.com/sosy-lab/sv-benchmarks/blob/master/c/pthread-divine/one_time_barrier_twice_2t.i
- https://github.com/sosy-lab/sv-benchmarks/blob/master/c/pthread-divine/barrier_2t.i

rafaelsamenezes commented:
After running in svcomp20 tag:

Before:
- Correct True: 180
- Correct False: 784
- Incorrect True: 15
- Incorrect False: 8
- Score: 536

After:
- Correct True: 177 (-4)
- Correct False: 788 (+4)
- Incorrect True: 12 (-3)
- Incorrect False: 12 (+4)
- Score: 566 (+30)

I will merge this.

mikhailramalho commented:
This introduced several warnings during compilation, could you fix them @lucasccordeiro?

lucasccordeiro commented:
This PR should fix the warnings https://github.com/esbmc/esbmc/pull/203/.
